numpy/core/src/multiarray/item_selection.c

```diff
@@ -844,26 +848,47 @@ PyArray_Repeat(PyArrayObject *aop, PyObject *op, int axis)
     for (i = 0; i < axis; i++) {
         n_outer *= PyArray_DIMS(aop)[i];
     }
+
+    if (needs_refcounting) {
+        if (PyArray_GetDTypeTransferFunction(
+                1, chunk, chunk, PyArray_DESCR(aop), PyArray_DESCR(aop), 0,
+                &cast_info, &flags) < 0) {
+            goto fail;
+        }
+    }
+
     for (i = 0; i < n_outer; i++) {
         for (j = 0; j < n; j++) {
             npy_intp tmp = broadcast ? counts[0] : counts[j];
             for (k = 0; k < tmp; k++) {
-                memcpy(new_data, old_data, chunk);
+                if (!needs_refcounting) {
+                    memcpy(new_data, old_data, chunk);
+                }
+                else {
+                    char *data[2] = {old_data, new_data};
+                    npy_intp strides[2] = {chunk, chunk};
+                    npy_intp one = 1;
+                    if (cast_info.func(&cast_info.context, data, &one, strides,
```

Comment:
```suggestion
                    if (cast_info.func(&cast_info.context, data, &tmp, strides,
```
I think it makes sense to reorganize things to call this with `&tmp` here.

Moving things into a `repeat_impl` helper that includes the loops would allow to do the specialization trick for the typical "`chunk`" sizes (i.e. itemsize) via `if (chunk == 8) repeat_impl(...)`.
That would make your integer example much faster.

I admit that would be an unrelated improvement.  The change looks good, though.  I would be happy about benchmarks for repeat (we have none), but don't care too much.